Fix race between replica reset and primary promotion

ywelsch opened the pull request:
We've recently seen a number of test failures that tripped an assertion in IndexShard (see issues linked below), leading to the discovery of a race between resetting a replica when it learns about a higher term and when the same replica is promoted to primary. This PR fixes the race by distinguishing between a cluster state primary term (called pendingPrimaryTerm) and a shard-level operation term. The former is set during the cluster state update or when a replica learns about a new primary. The latter is only incremented under the operation block, which can happen in a delayed fashion. It also solves the issue where a replica that's still adjusting to the new term receives a cluster state update that promotes it to primary, which can happen in the situation of multiple nodes being shut down in short succession. In that case, the cluster state update thread would call `asyncBlockOperations` in `updateShardState`, which in turn would throw an exception as blocking permits is not allowed while an ongoing block is in place, subsequently failing the shard. This PR therefore extends the IndexShardOperationPermits to allow it to queue multiple blocks (which will all take precedence over operations acquiring permits). Finally, it also moves the primary activation of the replication tracker under the operation block, so that the actual transition to primary only happens under the operation block.

Relates to #32431, #32304 and #32118

bleskes commented on server/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java:
```diff
@@ -182,10 +179,14 @@ private void delayOperations() {
     private void releaseDelayedOperations() {
         final List<DelayedOperation> queuedActions;
         synchronized (this) {
```
I think we can pull this up to the method. I don't see a reason to drain the queue and the release the queue and it will simplify the reasoning a bit.

ywelsch replied:
I'm not sure what you mean. What would you change?

bleskes replied:
I updated my comment. I mean make this entire method synchronized.

ywelsch replied:
I find the reasoning simpler here if we don't extend the mutex to a section of the code which it does not need to cover. Are you ok keeping it as is?

bleskes replied:
yes, I'm ok. It's subjective.

bleskes commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -2216,10 +2218,11 @@ public void acquireReplicaOperationPermit(final long operationPrimaryTerm, final
         verifyReplicationTarget();
-        final boolean globalCheckpointUpdated;
         if (operationPrimaryTerm > primaryTerm) {
             synchronized (primaryTermMutex) {
```
I wonder if we can remove this and only lock `mutex` on this level (it's always good to avoid multiple locks if possible).

ywelsch replied:
I'm fine relaxing it and using mutex as we now use asyncBlockOperations. This means that in practice, we will have at most have the number of indexing threads block on this (while possibly a concurrent cluster state update comes in, trying to acquire mutex as well). The first indexing thread will increase `pendingPrimaryTerm`, and all the other ones that are blocked on mutex will just acquire mutex and do a quick noop. All subsequent writes will not acquire the mutex anymore as they will bypass the pre-flight check.

bleskes commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -2229,38 +2232,41 @@ public void acquireReplicaOperationPermit(final long operationPrimaryTerm, final
-                                "shard term already update.  op term [" + operationPrimaryTerm + "], shardTerm [" + primaryTerm + "]";
+
+                    synchronized (mutex) {
+                        final CountDownLatch termUpdated = new CountDownLatch(1);
```
I wonder if we should have a method called "setPrimaryTerm" which gets a primary + a runnable to run underly the async block. That method will be called both from here and from updateShardState and make sure that the semantics of the exposing the primary term (after submitting  async block and asserting we're under a mutex via assertions) are the same.

ywelsch replied:
I've given this a try in 70262d79b9e

ywelsch commented:
The CI failure on this highlighted another issue. When a replica is promoted to primary while there is still an ongoing replica operation on the shard, the operation can incorrectly use the primary term of the new shard. In the typical case, this is fortunately caught by the translog due to an extra term check there, but in the presence of a generation rollover at the wrong moment, even that might not hold. The issue is that the term is not incremented under the operation block. To solve this, I've pushed a102ef9 which distinguishes between the primary term that the shard is supposed to have because of the cluster state, and the primary term that is used by operations and by the underlying engine. The former is updated right away, whereas the latter is only updated under the operation block, ensuring that each operation with a permit always sees the correct term. Let me know what you think.

ywelsch commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -2216,10 +2218,11 @@ public void acquireReplicaOperationPermit(final long operationPrimaryTerm, final
         if (operationPrimaryTerm > primaryTerm) {
             synchronized (primaryTermMutex) {
                 if (operationPrimaryTerm > primaryTerm) {
+                    verifyNotClosed();
```
I did not *have* to (i.e. no failing test). I just saw that we were not rechecking this condition after possibly waiting for a while on `primaryTermMutex`. The next check 2 lines below will also fail this with an `IndexShardNotStartedException`. I found it nicer though to throw the `IndexShardClosedException` if possible.

ywelsch replied:
after second thought, this is less of an issue after I converted blockOperations to asyncBlockOperations in acquireReplicaOperationPermit. I'm going to revert

ywelsch commented:
@elasticmachine retest this please.

ywelsch commented:
@bleskes this is ready for another review. The main change (beside addressing your comments) I've done after the first review iteration was to better handle the primary activation of the replication tracker. Before the last iteration, the activation was done on the cluster state update thread, not when the shard actually became primary (i.e. under the operation block). This led to some test failures which are *fingers crossed* all fixed by the latest iteration.

bleskes commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -192,7 +193,8 @@
     protected volatile ShardRouting shardRouting;
     protected volatile IndexShardState state;
-    protected volatile long primaryTerm;
+    protected volatile long pendingPrimaryTerm;
```
can we have a comment pointing people to the java docs of `getPendingPrimaryTerm` for explanation of what it means?

ywelsch replied:
fixed in 4b82ca7f770

bleskes commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -2229,64 +2267,60 @@ public void acquireReplicaOperationPermit(final long operationPrimaryTerm, final
+                        bumpPrimaryTerm(opPrimaryTerm, () -> {
+                            // a primary promotion, or another primary term transition, might have been triggered concurrently to this
+                            // recheck under the operation permit if we can skip doing this work
+                            if (opPrimaryTerm == pendingPrimaryTerm) {
```
how is that possible? shouldn't the pending primary term update under mutex prevent this?

ywelsch replied:
I'm not sure how it can "prevent" this.

Assume you have a replica term bump followed by a promotion to primary. The replica term bump will call asyncBlockOperations to run the above code. Assume that acquireReplicaOperationPermit leaves the mutex before the code in the operation block gets to execute. Then a primary promotion comes in from the cluster state, updating the pendingPrimaryTerm again. Note that there's a test for this (`testReplicaTermIncrementWithConcurrentPrimaryPromotion`).

bleskes replied:
+1 to the test.

So say the increase of the term comes in first. That acquires the mutex, bumps the `pendingPrimaryTerm` and submits an async block under mutex.

Then the replica gets promoted. That means the term is higher. When the cluster state comes in, it bumps the `pendingPrimaryTerm` again and submit an async block. We know that they first async block code will run first, followed by the second, and I think that's OK.

If the reverse happens - i.e., the updateShardState comes in first, it will bump the pendingPrimaryTerm to a higher number than the replication operation, which will prevent the replication operation from submitting it's async block, so we're good here too. 

What am I missing?

ywelsch replied:
> We know that they first async block code will run first?

No we don't know that. See implementation of asyncBlockOperations. Both just submit a task to the generic threadpool. Both potentially race to the call of `doBlockOperations`

bleskes commented on server/src/main/java/org/elasticsearch/index/shard/StoreRecovery.java:
```diff
@@ -458,11 +459,11 @@ private void restore(final IndexShard indexShard, final Repository repository, f
             assert indexShard.shardRouting.primary() : "only primary shards can recover from store";
             indexShard.openEngineAndRecoverFromTranslog();
-            indexShard.getEngine().fillSeqNoGaps(indexShard.getPrimaryTerm());
+            indexShard.getEngine().fillSeqNoGaps(indexShard.getPendingPrimaryTerm());
```
This is another wart...  I wonder if we should fold it into openEngineAndRecoverFromTranslog  (another change)

bleskes commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -468,12 +475,20 @@ public void updateShardState(final ShardRouting newRouting,
                             try {
+                                synchronized (mutex) {
+                                    assert shardRouting.primary();
+                                    // do these updates under the mutex as this otherwise races with subsequent calls of updateShardState
```
what race conditions do refer to?

ywelsch replied:
I think this is obsolete and does not need to be done under the mutex anymore. It came from an earlier iteration where I had not introduced the `relocated` state in ReplicationTracker yet, and where I was using both `replicationTracker.isPrimaryMode()` and comparing `operationPrimaryTerm` to `pendingPrimaryTerm` to figure out if a shard had possibly relocated and wanted this to be an atomic thing.

ywelsch replied:
One problematic place I see is the assertion at the end of updateShardState which checks both isPrimaryMode and the operationTerm + pendingPrimaryTerm. If we don't update the replication tracker + the operationPrimaryTerm atomically under the mutex, this invariant might be violated.

bleskes commented:
LGTM, hopefully the CI failure is nothing challenging. Thanks for the extra push @ywelsch

ywelsch commented:
It's a BWC test that fails due to serialization changes. I'll merge latest master and try again.

ywelsch commented:
Thanks @bleskes

liaoyanyunde commented:
` private <E extends Exception> void bumpPrimaryTerm(long newPrimaryTerm, final CheckedRunnable<E> onBlocked) {
        assert Thread.holdsLock(mutex);
        assert newPrimaryTerm > pendingPrimaryTerm;
        assert operationPrimaryTerm <= pendingPrimaryTerm;`

Is it possible that newPrimaryTerm < pendingPrimaryTerm,and leader to bumpPrimaryTerm throw  an assert failures?

liaoyanyunde commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -468,12 +475,15 @@ public void updateShardState(final ShardRouting newRouting,
+                    bumpPrimaryTerm(newPrimaryTerm,
                         () -> {
                             shardStateUpdated.await();
+                            assert pendingPrimaryTerm == newPrimaryTerm :
```
1、in method updateShardState ,because newPrimaryTerm != pendingPrimaryTerm so the code run into bumpPrimaryTerm method 
![image](https://user-images.githubusercontent.com/44155914/65228772-a917c880-dafd-11e9-80d4-d1f44c7f6c33.png)

2、in bumpPrimaryTerm，it will set pendingPrimaryTerm = newPrimaryTerm
![image](https://user-images.githubusercontent.com/44155914/65228895-e419fc00-dafd-11e9-9177-620315ec849e.png)

3、but  before bumpPrimaryTerm set pendingPrimaryTerm = newPrimaryTerm ，onBlocked will run first，in which will assert pendingPrimaryTerm == newPrimaryTerm
![image](https://user-images.githubusercontent.com/44155914/65229035-22afb680-dafe-11e9-9f3e-91749a87da57.png)
![image](https://user-images.githubusercontent.com/44155914/65229109-470b9300-dafe-11e9-9683-7d8a607fd774.png)

liaoyanyunde commented:
You said that :
" It also solves the issue where a replica that's still adjusting to the new term receives a cluster state update that promotes it to primary, which can happen in the situation of multiple nodes being shut down in short succession. In that case, the cluster state update thread would call asyncBlockOperations in updateShardState, which in turn would throw an exception as blocking permits is not allowed while an ongoing block is in place, subsequently failing the shard"

Can you tell me more about this?

liaoyanyunde commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -468,12 +475,15 @@ public void updateShardState(final ShardRouting newRouting,
+                            assert pendingPrimaryTerm == newPrimaryTerm :
+                                "shard term changed on primary. expected [" + newPrimaryTerm + "] but was [" + pendingPrimaryTerm + "]" +
+                                ", current routing: " + currentRouting + ", new routing: " + newRouting;
+                            assert operationPrimaryTerm == newPrimaryTerm;
```
Is "assert operationPrimaryTerm == newPrimaryTerm;"  unnecessary here.

because in bumpPrimaryTerm it will set  operationPrimaryTerm == newPrimaryTerm before goto run onBlocked,so it seems that "assert operationPrimaryTerm == newPrimaryTerm;" will always be true

ywelsch commented:
@liaoyanyunde can you provide some context for these questions? Are you studying this code to understand how it works? To which goal?

> when acquireReplicaOperationPermit alls bumpPrimaryTerm , is it possible that newPrimaryTerm < pendingPrimaryTerm,and lead to bumpPrimaryTerm throw an assert failures?

No, primary terms are always non-decreasing, which is guaranteed by the cluster coordination subsystem.

> but before bumpPrimaryTerm set pendingPrimaryTerm = newPrimaryTerm ，onBlocked will run first，in which will assert pendingPrimaryTerm == newPrimaryTerm

The CountDownLatch `termUpdated` makes sure this scenario can't happen.

> Is it reasonable? it seems that the assertion "pendingPrimaryTerm == newPrimaryTerm" will always be false?

We run all our tests with assertions enabled. If this would not hold true, we would quickly learn about it.

> Can you tell me more about this?How is it to avoid failing the shard problem?

I'm not sure on what you seek clarification. Also, if this is purely about understanding the current shape of the code, perhaps it's not necessary to understand the full history on how we arrived here.

> Is "assert operationPrimaryTerm == newPrimaryTerm;" unnecessary here.

I'm not sure what you mean by "unnecessary". Assertions are there to validate our assumptions about the code. They should obviously always hold true.

liaoyanyunde commented:
> can you provide some context for these questions? Are you studying this code to understand how it works? To which goal?
We are incorporating the fix code for this issue into version 6.3.2. So I am researching this code to see how it works. But in the process of reading, there are many doubts.
